### ISSUE TITLE
Can we improve jaggies on UI meshes please?

### ISSUE THREAD

**Issue by Jumbli** [StereoKit/StereoKit#483]
## Describe the feature
UI elements have lots of straight lines and suffer greatly with jaggies. In my more recent videos I have been bumping up my resolution using OpenXR toolkit (for PC) and SideQuest for Quest to try and make the UI look better, but increasing the resolution has a performance cost and I can't ask end users to use external tools to control the resolution in the released app.

There may be multiple ways to progress with this:
1) Add antialiasing - but I assume this in not a trivial thing.
2) Add resolution scaling to StereoKit so we can render higher resolution and then down sample (I would welcome this regardless of antialiasing)

I have switched to using flat 2d surfaces for my windows and buttons because I think that reduces artifacts a little. If the above 2 options require a lot of work, an alternative feature that may help me is to allow UV maps to be defined for custom meshes, I could then try fading out the hard edge of UI surfaces in my custom UI quadrant shader.

Thank you.

**Comment by maluoi** [comment]
Yeah, I can start looking into this! Are you interested in this for Quest standalone, Quest+Link, or both? I know Quest standalone has a lamentable amount of aliasing, but desktop tends to be mostly okay. I'd be interested in your opinion on that though!

One thing is that the current UI mesh is somewhat prone to aliasing even in the best cases. The colored edges quite frequently end up at glancing angles, creating high-contrast slivers that sparkle! I've been experimenting with some edges that wrap-around the corners to prevent those slivers, but I haven't had the time to really dig into a cohesive design using it.

![image](https://user-images.githubusercontent.com/8484884/194376144-4b57e08e-a31d-4626-a129-b5a0ae4e8d8d.png)

If you're interested in trying it out or playing around with it, the code's over here:
https://github.com/StereoKit/StereoKit/blob/develop/Examples/StereoKitTest/Tools/HolographicTheme.cs#L54

**Comment by Jumbli** [comment]
Thanks for that. I'm currently targeting Quest and PC, but basically want to get on every device I can.

While in the experience on PC, it didn't bother me too much. It's when I came to do videos (based on the PC version) that it stood out much more and that's when I started overriding the default resolution. A problem is that after release, reviewers will upload their own footage to YouTube at the default resolution and that will highlight the aliasing issues on the UI and could put people off, so for marketing purposes it's important to have the highest quality visuals I can.

I've already been adapting the Holographic theme for my flat style which has the benefit of getting rid of the depth and high contrast edges. The following is from the Quest. For buttons I have no edge but keep a 2d drop shadow. The drop shadow is prone to jaggies so it would be even better if I could reduce the transparency for just that element. Up close it does feel better, but as you go further away you notice the aliasing much more. This is from a Quest screenshot.

![Quest](https://user-images.githubusercontent.com/10079865/194392802-944965e5-a1fc-49c8-8848-d5468bc1bd56.jpg)

**Comment by Jumbli** [comment]
Also, text usually looks bad in VR when it's on a distant UI so I will probably experiment with doing a depth fade on the font to reduce that issue.

**Comment by Jumbli** [comment]
I've done some work on adding a fade around the edges of buttons and windows to reduce jaggies and to fade text at a distance. It seems to make a significant improvement. There is an overdraw overhead due to having them transparent, but performance seems OK so far.

It's worth noting that it's common practice for Quest games (and Rift) to increase the resolution to maybe 1.3 pixel density and this makes a good improvement to the general clarity of Quest software. I wonder if there is an OpenXR solution in the works for resolution scaling. Antialiasing would still be welcome when you have the opportunity :)

https://user-images.githubusercontent.com/10079865/194718867-a6fc0067-0f3a-4f9d-bd10-7c70dec82d83.mp4

**Comment by maluoi** [comment]
Scaling the display texture seems to be pretty trivial to add, so I'll definitely be shipping that! It does make a pretty dramatic difference on Quest. I still want to resolve some thoughts about how to determine good per-platform defaults with this, but I don't think that'll prevent me from shipping the basic feature.

![StereoKitResolutionSettings](https://user-images.githubusercontent.com/8484884/194720381-cb69a688-737b-4d73-ae73-410737a4e712.gif)

MSAA is proving a little more complicated. You can see it working well enough in the gif above taken in the WMR runtime, but I've found some issues I still need to explore. Might still be on my code yet, but when I ran this on the Oculus desktop runtime, enabling MSAA blanked out the right eye. Still need to get to the GL side as well.

Your button fading is pretty awesome, that does look really good like that! I've been _trying_ to stay away from transparency to keep rendering simpler, but that is pretty tempting. I still haven't given up hope on text clarity though, I think there's still some work that can be done there to improve that :D

**Comment by Jumbli** [comment]
Wow, that was quick progress. I would recommend defaulting to the native resolution for the device (or whatever OpenXR tells you). It's then up to the developer to decide what to set as the default for each device. They know how hard they are pushing the hardware and what a suitable default should be, For PC they often allow users to change the setting in the app too.

For text clarity switching to SDFs should make a massive difference. I think the open source font library you currently use already has some code for SDF support, but I haven't looked too closely.

I'm looking forward to the render scaling so I can publish some native Quest videos :) Thanks.

**Comment by maluoi** [comment]
This is now in v0.3.7-preview.8! I've found it to be a little unsupported by some OpenXR runtimes. I haven't determined if it's in my code or the runtime's code, but your mileage may vary. Among other things, Oculus seems to ask for the app not to use MSAA, which is a little odd, and some runtimes only display one eye. I'd consider this feature a bit of a WIP.

Render scaling should be pretty reliable though!

**Comment by Jumbli** [comment]
Thanks for those changes.

Quest 1 & 2
renderScaling works fine :)
renderMultisample is being ignored. I found the following forum post which indicates developers are just ignoring the maxSwapchainSampleCount and assume it's a bug that it returns 1. In fact, Meta do recommend that Quest apps use MSAA so can we do the same as other devs for Quest and allow it to be overridden to 2 or 4 regardless of maxSwapchainSampleCount?
(https://communityforums.atmeta.com/t5/OpenXR-Development/maxSwapchainSampleCount-is-1-on-Quest-2-No-multisample-swapchain/td-p/893389)

Rift
renderScaling distorts the screen making it unusable.
renderMultisample returns the following error and makes the screen black,
SK info] xrEndFrame [XR_ERROR_RUNTIME_FAILURE]

Thanks.

**Comment by maluoi** [comment]
Yeah, this is something I ran into and wasn't certain what exactly to do with it! I'd prefer to obey the `maxSwapchainSampleCount` to keep the code portable, but yeah, I've also seen their docs where they basically insist on MSAA! (yay) So it may be time to break out some per-runtime exceptions Thanks for the thread there too, I chimed in on it as well! Would be nice to get that fixed at the source instead.

I wasn't actually sure if I should release this feature or not, given how many runtimes seem to have some amount of undefined behavior related to it! The one eye issue, or renderScaling zoom seems to happen on some platforms, but not others. I think ultimately what I'll need to do is stop relying on OpenXR for anything other than the most basic of presentation paths (whatever that may end up being), and do all of that functionality internal to StereoKit. It's a good chunk of work, so in the meantime I'd consider this a setting that you should define per-platform after testing it out. Not ideal, but I figured it'd still be helpful for you in the cases where it does work.

**Comment by Jumbli** [comment]
That's great. Thanks.

Quest is my main worry at the moment, so getting that multi-sample working will be a good win.
I'm fine with it being a "developer beware" feature and it being up to the developer to prove it works on their supported platforms and excluding it when unsure.

**Comment by maluoi** [comment]
#703 will have code in it that should help decouple MSAA from OpenXR, and will allow for more universally controllable MSAA settings.
